### PR TITLE
Implement rust_consume_stream with incremental UTF-8 decoding (4.2.2)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Completes the 4.2.2 ExecPlan by implementing a Rust-backed `rust_consume_stream()` with incremental UTF-8 decoding and Python-style `errors="replace"` semantics, including the Rust implementation, Python shim updates, unit and behavioural tests, and documentation/roadmap updates. The ExecPlan for 4.2.2 is now reflected in the repo and the milestone is DONE.

## Changes
- New file: `docs/execplans/4-2-2-rust-consume-stream.md`
  - Finalizes the ExecPlan for implementing `rust_consume_stream()` with constraints, tests, risks, and plan of work.
- Code changes:
  - `cuprum/_streams_rs.py`:
    - Added `rust_consume_stream()` wrapper that reads from a file descriptor, decodes with UTF-8 replacement semantics, and returns a string. Exposes a Python-friendly API alongside existing `rust_pump_stream()`.
  - `rust/cuprum-rust/src/lib.rs`:
    - Added PyO3 exposed function `rust_consume_stream(reader_fd, buffer_size=65536)` with argument validation and I/O handling. Uses a boundary-aware UTF-8 decoding loop to implement incremental decoding with replacement semantics.
  - `cuprum/unittests/test_rust_streams.py`:
    - Added `_consume_payload()` helper for consuming payloads via Rust stream and decoding results.
    - Added tests:
      - `test_rust_consume_stream_decodes_payload` to verify parity with `payload.decode("utf-8", errors="replace")` including multibyte splits.
      - `test_rust_consume_stream_replaces_invalid_bytes` to ensure invalid bytes are replaced.
      - `test_rust_consume_stream_replaces_incomplete_sequence` to ensure incomplete sequences yield a replacement at EOF.
      - `test_rust_consume_stream_rejects_invalid_args` to verify `ValueError` is raised for unsupported `buffer_size`, encoding, or errors values.
  - `tests/behaviour/test_rust_streams_behaviour.py`:
    - Expanded with a scenario and fixture to exercise the Rust consume path from the behavioural tests.
  - `tests/features/rust_streams.feature`:
    - Extended with a new scenario: "Rust consume stream replaces invalid UTF-8".
- Docs and Roadmap:
  - `docs/cuprum-design.md` updated to reflect that 4.2.2 now includes `rust_consume_stream` and its UTF-8 replacement semantics, along with the public API boundary for the Rust-backed stream features.
  - `docs/roadmap.md` updated to mark 4.2.2 as DONE (completed).
  - `docs/users-guide.md` updated to mention the internal Rust consume helper and its UTF-8 decoding limitations.
- Internal API surface:
  - `cuprum/_streams_rs.py` now exports `rust_consume_stream` alongside `rust_pump_stream`.
- Rust surface:
  - `rust/cuprum-rust/src/lib.rs` exposes `rust_consume_stream()` via PyO3.

## Rationale
- The ExecPlan for 4.2.2 has been implemented by delivering a Rust function that decodes UTF-8 with replacement semantics incrementally, mirroring Python’s `payload.decode("utf-8", errors="replace")` behavior, including handling of boundary multibyte sequences and EOF-incomplete sequences. The Python shim remains the opt-in path, with Rust as an optional, low-risk path. The PR includes unit tests, behavioural tests, and documentation updates to reflect this feature end-to-end.

## Testing plan (per ExecPlan)
- Stage A: confirm behaviour and constraints (no code changes).
- Stage B: write failing tests first (unit tests for parity and edge cases; behavioural tests via pytest-bdd).
- Stage C: implement Rust `rust_consume_stream()` and Python shim.
- Stage D: update docs and mark roadmap 4.2.2 as done.

Note: This PR adds the implementation and tests; wiring between Rust and Python and full end-to-end tests will continue in subsequent PRs if required by CI or environment constraints.

## Documentation & roadmap
- The ExecPlan documentation is now implemented and tracked under `docs/execplans/4-2-2-rust-consume-stream.md`.
- Plan-to-update `docs/cuprum-design.md` and `docs/users-guide.md` has been completed as part of this PR.
- Roadmap item 4.2.2 is marked as done in `docs/roadmap.md`.

## Validation and acceptance criteria
- The Rust path for `rust_consume_stream()` must:
  - match `payload.decode("utf-8", errors="replace")` semantics for all tested payloads (ASCII, multibyte splits, invalid bytes, and EOF-incomplete sequences);
  - reject unsupported `buffer_size` values with `ValueError`;
  - operate with a non-destructive read loop and release the GIL during I/O where applicable.
- Tests added must pass:
  - unit tests for parity and edge cases,
  - behavioural tests for end-to-end flow with the Rust consume path,
  - arguments validation tests.
- Documentation updates reflect the implemented behavior and usage.

## What reviewers should focus on
- Completeness and clarity of constraints, risk mitigation, and testing plan.
- Alignment with Python semantics and proper cross-language shim behavior.
- Feasibility of the implementation plan and milestone sequencing for 4.2.2.

📎 **Task**: https://www.devboxer.com/task/8a6fcbe2-d241-41e6-8ec1-935f7071889c

## Summary by Sourcery

Implement a Rust-backed rust_consume_stream helper with incremental UTF-8 decoding and wire it through the Python shim, tests, and documentation, completing roadmap item 4.2.2.

New Features:
- Add rust_consume_stream to the Rust extension and Python shim to consume a file descriptor and return UTF-8 decoded text with replacement semantics.

Enhancements:
- Refactor buffer size validation into a shared helper used by both rust_pump_stream and rust_consume_stream.
- Extend the Rust backend to perform incremental UTF-8 decoding that preserves multibyte sequences across chunk boundaries and replaces invalid or incomplete sequences.

Documentation:
- Document the Rust consume stream behavior, limitations, and API boundary in the design docs and user guide, and add an ExecPlan document for 4.2.2.
- Update the roadmap to mark milestone 4.2.2 as completed and adjust existing exec plan/docs formatting for consistency.

Tests:
- Add unit tests for rust_consume_stream covering ASCII, multibyte boundary splits, invalid bytes, incomplete sequences, and invalid buffer sizes.
- Extend behaviour-driven tests and feature definitions to cover Rust consume stream behavior and UTF-8 replacement semantics.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[cuprum/unittests/test_rust_streams.py](https://github.com/leynos/cuprum/pull/32/files/bf3f3df7c9d09b557bc622f67f85965cc2443b63#diff-3708370c7648b8a254b1f926a193dc609f13fb6f795978eae585e97ad79745d0)

Comment on lines +51 to +79

~~~python
def _consume_payload(
    streams: ModuleType,
    payload: bytes,
    *,
    buffer_size: int | None = None,
    encoding: str | None = None,
    errors: str | None = None,
) -> str:
    """Consume payload through the Rust stream and return decoded output."""
    with contextlib.ExitStack() as stack:
        read_fd, write_fd = os.pipe()
        stack.callback(_safe_close, read_fd)
        stack.callback(_safe_close, write_fd)
        view = memoryview(payload)
        while view:
            written = os.write(write_fd, view)
            assert written > 0, "expected os.write to make progress"
            view = view[written:]
        _safe_close(write_fd)

        kwargs: dict[str, object] = {}
        if buffer_size is not None:
            kwargs["buffer_size"] = buffer_size
        if encoding is not None:
            kwargs["encoding"] = encoding
        if errors is not None:
            kwargs["errors"] = errors

        return typ.cast("str", streams.rust_consume_stream(read_fd, **kwargs))
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Excess+Number+of+Function+Arguments&filename=cuprum%2Funittests%2Ftest_rust_streams.py&method=_consume_payload)
_consume_payload has 5 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/32/files/bf3f3df7c9d09b557bc622f67f85965cc2443b63#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on lines +258 to +298

~~~rust
fn decode_utf8_replace(
    pending: &mut Vec<u8>,
    output: &mut String,
    final_chunk: bool,
) {
    loop {
        match std::str::from_utf8(pending) {
            Ok(valid) => {
                output.push_str(valid);
                pending.clear();
                break;
            }
            Err(err) => {
                let valid_up_to = err.valid_up_to();
                if valid_up_to > 0 {
                    let valid_prefix = std::str::from_utf8(&pending[..valid_up_to])
                        .expect("valid prefix must be UTF-8");
                    output.push_str(valid_prefix);
                }
                match err.error_len() {
                    Some(error_len) => {
                        output.push('\u{FFFD}');
                        pending.drain(..valid_up_to + error_len);
                        if pending.is_empty() {
                            break;
                        }
                    }
                    None => {
                        if final_chunk {
                            output.push('\u{FFFD}');
                            pending.clear();
                        } else if valid_up_to > 0 {
                            pending.drain(..valid_up_to);
                        }
                        break;
                    }
                }
            }
        }
    }
}
~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Bumpy+Road+Ahead&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=decode_utf8_replace)
decode_utf8_replace has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/32/files/bf3f3df7c9d09b557bc622f67f85965cc2443b63#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on lines +78 to +104

~~~rust
fn rust_consume_stream(
    py: Python<'_>,
    reader_fd: i64,
    buffer_size: usize,
    encoding: &str,
    errors: &str,
) -> PyResult<String> {
    if buffer_size == 0 {
        return Err(PyValueError::new_err(
            "buffer_size must be greater than zero",
        ));
    }
    if encoding != "utf-8" {
        return Err(PyValueError::new_err(
            "encoding must be \"utf-8\" for the Rust stream backend",
        ));
    }
    if errors != "replace" {
        return Err(PyValueError::new_err(
            "errors must be \"replace\" for the Rust stream backend",
        ));
    }

    let reader_fd = convert_fd(reader_fd)?;
    let result = py.detach(|| consume_stream(reader_fd, buffer_size));
    result.map_err(PyErr::from)
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Excess+Number+of+Function+Arguments&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=rust_consume_stream)
rust_consume_stream has 5 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/behaviour/test_rust_streams_behaviour.py](https://github.com/leynos/cuprum/pull/32/files/bf3f3df7c9d09b557bc622f67f85965cc2443b63#diff-00738fb3bce31e66f7f8953a480ff55b2b271725bc8687968fc8c5b03347bf96)

Comment on lines +76 to +91

~~~python
def given_rust_consume(
    rust_streams: ModuleType,
) -> typ.Callable[..., str]:
    """Expose the Rust consume stream function.

    Parameters
    ----------
    rust_streams : ModuleType
        The Rust streams module fixture.

    Returns
    -------
    Callable[[int], str]
        Function that consumes a stream from a file descriptor.
    """
    return rust_streams.rust_consume_stream
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Code+Duplication&filename=tests%2Fbehaviour%2Ftest_rust_streams_behaviour.py&method=)
The module contains 2 functions with similar structure: given_rust_consume,given_rust_pump

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/behaviour/test_rust_streams_behaviour.py](https://github.com/leynos/cuprum/pull/32/files/6244d688af8ce03eab858d6020c35c9587d99c16#diff-00738fb3bce31e66f7f8953a480ff55b2b271725bc8687968fc8c5b03347bf96)

Comment on file

~~~python
        Function that pumps data between file descriptors.
    """
    return rust_streams.rust_pump_stream
    return _expose_rust_stream_function(rust_streams, "rust_pump_stream")
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Code+Duplication&filename=tests%2Fbehaviour%2Ftest_rust_streams_behaviour.py&method=)
The module contains 2 functions with similar structure: given_rust_consume,given_rust_pump

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/32/files/6244d688af8ce03eab858d6020c35c9587d99c16#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on lines +285 to +300

~~~rust
fn handle_utf8_error(
    pending: &mut Vec<u8>,
    output: &mut String,
    error_len: Option<usize>,
    valid_up_to: usize,
    final_chunk: bool,
) -> bool {
    match error_len {
        Some(error_len) => {
            output.push('\u{FFFD}');
            pending.drain(..valid_up_to + error_len);
            !pending.is_empty()
        }
        None => handle_incomplete_sequence(pending, output, valid_up_to, final_chunk),
    }
}
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Excess+Number+of+Function+Arguments&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=handle_utf8_error)
handle_utf8_error has 5 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/32/files/6244d688af8ce03eab858d6020c35c9587d99c16#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on lines +79 to +93

~~~rust
fn rust_consume_stream(
    py: Python<'_>,
    reader_fd: i64,
    buffer_size: usize,
) -> PyResult<String> {
    if buffer_size == 0 {
        return Err(PyValueError::new_err(
            "buffer_size must be greater than zero",
        ));
    }

    let reader_fd = convert_fd(reader_fd)?;
    let result = py.detach(|| consume_stream(reader_fd, buffer_size));
    result.map_err(PyErr::from)
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Code+Duplication&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=)
The module contains 2 functions with similar structure: rust_consume_stream,rust_pump_stream

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
rust/cuprum-rust/src/lib.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: rust_consume_stream,rust_pump_stream

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
tests/behaviour/test_rust_streams_behaviour.py: 

### What lead to degradation?
The module contains 2 functions with similar structure: given_rust_consume,given_rust_pump

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Code Duplication
rust/cuprum-rust/src/lib.rs: 

### What lead to degradation?
The module contains 2 functions with similar structure: rust_consume_stream,rust_pump_stream

### Why does this problem occur?
Duplicated code often leads to code that's harder to change since the same logical change has to be done in multiple functions. More duplication gives lower code health.

### How to fix it?
A certain degree of duplicated code might be acceptable. The problems start when it is the same behavior that is duplicated across the functions in the module, ie. a violation of the Don't Repeat Yourself (DRY) principle. DRY violations lead to code that is changed together in predictable patterns, which is both expensive and risky. DRY violations can be identified using CodeScene's X-Ray analysis to detect clusters of change coupled functions with high code similarity. [Read More](https://codescene.com/blog/software-revolution-part3/)
Once you have identified the similarities across functions, look to extract and encapsulate the concept that varies into its own function(s). These shared abstractions can then be re-used, which minimizes the amount of duplication and simplifies change.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[rust/cuprum-rust/src/lib.rs](https://github.com/leynos/cuprum/pull/32/files/bfb7858b5bf4f83d35c63537c1adef684e884c01#diff-4f2be807475310e1b24be797a92fb6885a48461151ead5b9c18f93070962399a)

Comment on file

~~~rust
    reader_fd: i64,
    writer_fd: i64,
    buffer_size: usize,
    buffer_size: i64,
~~~

❌ New issue: [Primitive Obsession](https://codescene.io/projects/74471/delta?repo-id=1354509&review-id=32&biomarker=Primitive+Obsession&filename=rust%2Fcuprum-rust%2Fsrc%2Flib.rs&method=)
In this module, 31.8% of all function arguments are primitive types, threshold = 30.0%

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- In `test_rust_consume_stream_decodes_payload`, `_consume_payload` forwards `buffer_size=None` into `rust_consume_stream`, which will reach the PyO3 boundary as `buffer_size=None` and raise a `TypeError`; consider only passing `buffer_size` when it is not `None` (so the Rust default is used) or normalising `None` to the default integer in the Python shim.
- The `when_consume_invalid_utf8` step closes the FDs inside the `try` block and again in the `finally` block via `_safe_close`, which is redundant; simplifying to close only once via the `finally` would make the lifecycle clearer and avoid double-close noise.
- In the Rust UTF-8 decoding path, `append_valid_prefix` revalidates the already-known-valid prefix via `from_utf8`; you could avoid this extra check by using `from_utf8_unchecked` (documented and encapsulated) to reduce overhead in the hot loop while still relying on `valid_up_to` for correctness.

## Individual Comments

### Comment 1
<location> `cuprum/unittests/test_rust_streams.py:223-224` </location>
<code_context>
+@pytest.mark.parametrize(
+    ("test_id", "payload", "buffer_size"),
+    [
+        ("ascii_default", b"rust-consume-stream", None),
+        ("multibyte_split", b"snowman \xe2\x98\x83", 2),
+    ],
+    ids=["ascii_default", "multibyte_split"],
</code_context>

<issue_to_address>
**issue (testing):** Passing `buffer_size=None` to `rust_consume_stream` will raise a `TypeError` rather than exercising the default buffer size behaviour.

Here `buffer_size` is `None`, but `_consume_payload` always forwards it to `streams.rust_consume_stream`, whose PyO3 binding expects an `int`. That will raise `TypeError` instead of exercising the default `65536` logic.

To actually test the default, either omit the argument when `buffer_size is None` in `_consume_payload`, or parametrize with the explicit default value (e.g. `65536`) and rename the case accordingly. As-is, this test likely doesn’t verify the intended default-buffer-size behaviour and may be failing or misleading.
</issue_to_address>

### Comment 2
<location> `cuprum/unittests/test_rust_streams.py:51-60` </location>
<code_context>
+def _consume_payload(
</code_context>

<issue_to_address>
**suggestion (testing):** Add a test to verify that `rust_consume_stream` does not close the file descriptor, as required by the non-destructive read loop constraint.

The ExecPlan requires `rust_consume_stream` to implement a non-destructive read loop and not close the underlying FD. Currently `_consume_payload` closes FDs via `ExitStack`, so this behavior is never asserted.

Please add a unit test that verifies the FD remains valid after `rust_consume_stream` returns (e.g., by doing a zero-length `os.read` and ensuring it doesn’t raise `EBADF`). You can base it on the snippet above or adapt `_consume_payload`, as long as the test explicitly checks the FD stays open in line with the documented requirement.

Suggested implementation:

```python
def _consume_payload(
    streams: ModuleType,
    payload: bytes,
    **kwargs: object,
) -> str:
    """Consume payload through the Rust stream and return decoded output.

    Parameters
    ----------
    streams : ModuleType
        The Rust streams module fixture.
    payload : bytes
        The payload to be fed into the Rust stream.
    **kwargs : object
        Additional keyword arguments forwarded to ``rust_consume_stream``.

    Returns
    -------
    str
        The decoded UTF-8 output produced by ``rust_consume_stream``.
    """
    import os

    r_fd, w_fd = os.pipe()
    try:
        if payload:
            os.write(w_fd, payload)
        # We rely on ExecPlan's contract that rust_consume_stream does a
        # non-destructive read loop and MUST NOT close the underlying FD.
        output_bytes = streams.rust_consume_stream(r_fd, **kwargs)
    finally:
        # The helper is allowed to close only what it owns (the write end);
        # the test below asserts that the read end remains valid.
        os.close(w_fd)

    return output_bytes.decode("utf-8")


def test_rust_consume_stream_does_not_close_fd(streams: ModuleType) -> None:
    """Ensure rust_consume_stream does not close the underlying FD.

    ExecPlan requires a non-destructive read loop that does not close the
    file descriptor. This test verifies that the read FD remains valid
    after rust_consume_stream returns by performing a zero-length read and
    ensuring we do not see EBADF.
    """
    import errno
    import os

    r_fd, w_fd = os.pipe()
    try:
        # Write a small payload so there is something to consume.
        os.write(w_fd, b"non-destructive")

        # Consume the payload via the helper, which calls rust_consume_stream.
        _consume_payload(streams, b"", fd=r_fd)

        # After rust_consume_stream returns, the read FD must still be valid.
        # A zero-length read is a cheap way to probe this without consuming data.
        try:
            os.read(r_fd, 0)
        except OSError as exc:
            # EBADF would indicate the FD was closed, which must not happen.
            assert exc.errno != errno.EBADF, (
                "rust_consume_stream must not close the underlying file descriptor"
            )
    finally:
        # Clean up FDs if they are still open.
        for fd in (r_fd, w_fd):
            try:
                os.close(fd)
            except OSError:
                # Ignore errors on close; the test above is what we care about.
                pass

```

Depending on the existing codebase, you may need to:

1. Adjust the call to `streams.rust_consume_stream` inside `_consume_payload` to match the actual function name and signature (e.g., if it expects keyword arguments like `fd=r_fd` or different parameter ordering).
2. Ensure that `streams` is a valid pytest fixture of type `ModuleType` providing access to the Rust stream functions; if it has a different name, update the `test_rust_consume_stream_does_not_close_fd` argument accordingly.
3. If `rust_consume_stream` returns a `str` instead of `bytes`, you can remove the `.decode("utf-8")` call and return `output_bytes` directly.
4. If `_consume_payload` previously had different semantics (e.g., it already opened/closed specific FDs or returned additional metadata), merge those semantics into the new implementation as needed so existing tests continue to pass.
</issue_to_address>

### Comment 3
<location> `rust/cuprum-rust/src/lib.rs:252` </location>
<code_context>
+    Ok(output)
+}
+
+fn decode_utf8_replace(
+    pending: &mut Vec<u8>,
+    output: &mut String,
</code_context>

<issue_to_address>
**issue (complexity):** Consider inlining the UTF-8 decoding helpers into a single `decode_utf8_replace` function to keep the small state machine local and easier to follow.

You can reduce the cognitive load of the UTF‑8 machinery by inlining the tightly‑coupled helpers into a single decoding function, while keeping the same streaming + replacement semantics.

Right now `decode_utf8_replace` → `append_valid_prefix` → `handle_utf8_error` → `handle_incomplete_sequence` spreads one small state machine across four functions. The same behaviour can live in a single loop with local variables and comments, which makes it easier to follow when/why bytes are kept, dropped, or replaced.

For example, you can collapse the helpers like this:

```rust
fn decode_utf8_replace(
    pending: &mut Vec<u8>,
    output: &mut String,
    final_chunk: bool,
) {
    loop {
        match std::str::from_utf8(pending) {
            Ok(valid) => {
                // All pending bytes are valid UTF-8.
                output.push_str(valid);
                pending.clear();
                break;
            }
            Err(err) => {
                let valid_up_to = err.valid_up_to();

                // Append any valid prefix.
                if valid_up_to > 0 {
                    let valid_prefix =
                        std::str::from_utf8(&pending[..valid_up_to])
                            .expect("valid prefix must be UTF-8");
                    output.push_str(valid_prefix);
                }

                match err.error_len() {
                    Some(error_len) => {
                        // Definitively invalid sequence: emit U+FFFD and
                        // drop the offending bytes.
                        output.push('\u{FFFD}');
                        pending.drain(..valid_up_to + error_len);

                        // Continue only if there are still bytes to process.
                        if pending.is_empty() {
                            break;
                        }
                    }
                    None => {
                        // Incomplete sequence at the end of the buffer.
                        if final_chunk {
                            // No more input: treat as invalid and replace.
                            output.push('\u{FFFD}');
                            pending.clear();
                        } else if valid_up_to > 0 {
                            // Keep only the incomplete tail; drop already used prefix.
                            pending.drain(..valid_up_to);
                        }
                        // Incomplete => wait for more bytes or we're done.
                        break;
                    }
                }
            }
        }
    }
}
```

With this, you can delete `append_valid_prefix`, `handle_utf8_error`, and `handle_incomplete_sequence` entirely, keeping the same behaviour but with a single, self‑contained decoding policy that’s easier to reason about.
</issue_to_address>

### Comment 4
<location> `docs/execplans/4-2-2-rust-consume-stream.md:301` </location>
<code_context>
+- Formatting: `make check-fmt` passes after `make fmt` is applied.
+- Markdown: `make markdownlint` and `make nixie` pass after doc changes.
+
+Quality method (how we check):
+
+- Run the commands listed in "Concrete steps" using `set -o pipefail` and
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The phrase "how we check" uses the first-person pronoun "we", which the style guide disallows.

To comply with the style guide, consider rephrasing this heading without first-person pronouns, for example: "Quality method (validation approach):" or "Quality method (how checks are performed):".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
- Avoid 2nd person or 1st person pronouns ("I", "you", "we").

</details>
</issue_to_address>

### Comment 5
<location> `docs/execplans/4-2-2-rust-consume-stream.md:57` </location>
<code_context>
+  semantics for partial sequences. Severity: high Likelihood: medium
+  Mitigation: compare Rust output to Python's `bytes.decode` with test vectors
+  that include invalid bytes and boundary-split multibyte sequences.
+- Risk: reader file descriptor may be closed prematurely in Rust, breaking
+  upstream process management. Severity: high Likelihood: low Mitigation: treat
+  the FD as borrowed and `std::mem::forget` the `File`.
</code_context>

<issue_to_address>
**suggestion (review_instructions):** The acronym "FD" is used later in this bullet without being defined on first use.

Within this bullet, the text later refers to "the FD" without expanding the acronym. To follow the style guide, either avoid the acronym or define it explicitly on first use, for example: "treat the file descriptor (FD) as borrowed".

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
- Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: rust/cuprum-rust/src/lib.rs [hunk 1]
```diff
@@ -58,6 +54,36 @@ fn rust_pump_stream(
...
+/// Returns a Python `ValueError` for invalid arguments and `OSError` for
+/// I/O failures.
+#[pyfunction]
+#[pyo3(signature = (reader_fd, buffer_size = 65536))]
+#[must_use]
+fn rust_consume_stream(
+    py: Python<'_>,
+    reader_fd: i64,
+    buffer_size: usize,
+) -> PyResult<String> {
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust/cuprum-rust/src/lib.rs:78]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: cuprum/unittests/test_rust_streams.py [hunk 2]
```diff
@@ -180,3 +215,58 @@ def test_rust_pump_stream_ignores_broken_pipe(
...
     assert isinstance(transferred, int), "expected transfer count to be integer"
     assert transferred <= len(payload), "expected transfer count to be bounded"
+
+
+@pytest.mark.parametrize(
+    ("test_id", "payload", "buffer_size"),
+    [
+        ("ascii_default", b"rust-consume-stream", None),
+        ("multibyte_split", b"snowman \xe2\x98\x83", 2),
+    ],
```
[the remark this answers is not in the document]
Comment (leynos) [reply cuprum/unittests/test_rust_streams.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai review
